File: 2022/day21.cpp
// https://adventofcode.com/2022/day/21

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <functional>
#include <print>

using value_t = aoc::ratio<i64>;
using operation_t = std::function<value_t(value_t, value_t)>;

template <class index_t, class Operation = operation_t>
struct monkey_t {
  index_t index_lhs;
  index_t index_rhs;

  Option<value_t> value_lhs;
  Option<value_t> value_rhs;
  Operation operation;

  fn is_operation_ready() const -> bool {
    return value_lhs.has_value() && value_rhs.has_value();
  }
  fn apply_operation() const -> value_t {
    return operation(value_lhs.value(), value_rhs.value());
  }
};

using parsing_monkey_t = monkey_t<String, char>;
using parsed_monkeys_t = Vec<std::pair<String, parsing_monkey_t>>;
using monkeys_t = Vec<monkey_t<int>>;
using monkey_it = monkeys_t::iterator;

constexpr let value_op = '@';

fn parse(String const& filename) -> parsed_monkeys_t {
  auto parsed_monkeys = parsed_monkeys_t{};
  for (str line : aoc::views::read_lines(filename)) {
    let[name, full_op] = aoc::split_once<String>(line, ':');
    let[lhs, op, rhs] =
        aoc::split<std::array<String, 3>>(full_op.substr(1), ' ');

    if (op.empty()) {
      value_t value = aoc::to_number<int>(lhs);
      parsed_monkeys.emplace_back(name,
                                  parsing_monkey_t{"", "", value, 0, value_op});
      continue;
    }

    parsed_monkeys.emplace_back(name,
                                parsing_monkey_t{lhs, rhs, None, None, op[0]});
  }
  return parsed_monkeys;
}

fn get_value_monkey(value_t lhs, value_t rhs) -> monkey_t<int> {
  return {-1, -1, lhs, rhs, std::plus{}};
}

fn get_yelled_number(monkeys_t& monkeys, int current) -> value_t {
  auto& current_monkey = monkeys[current];
  if (current_monkey.is_operation_ready()) {
    return current_monkey.apply_operation();
  }
  current_monkey.value_lhs =
      get_yelled_number(monkeys, current_monkey.index_lhs);
  current_monkey.value_rhs =
      get_yelled_number(monkeys, current_monkey.index_rhs);
  return current_monkey.apply_operation();
}

fn transform_monkeys(parsed_monkeys_t const& parsed_monkeys)
    -> std::tuple<monkeys_t, int, int> {
  monkeys_t monkeys;
  let begin_it = std::begin(parsed_monkeys);
  int root_index = 0;
  int human_index = 0;
  for (int index = 0; let& [ name, monkey ] : parsed_monkeys) {
    if (name == "root") {
      root_index = index;
    } else if (name == "humn") {
      human_index = index;
    }
    ++index;

    if (monkey.operation == value_op) {
      monkeys.push_back(
          get_value_monkey(monkey.value_lhs.value(), monkey.value_rhs.value()));
      continue;
    }
    let lhs_it = stdr::find_if(parsed_monkeys, [&](let& monkey_pair) {
      return monkey_pair.first == monkey.index_lhs;
    });
    let rhs_it = stdr::find_if(parsed_monkeys, [&](let& monkey_pair) {
      return monkey_pair.first == monkey.index_rhs;
    });
    monkeys.emplace_back(std::distance(begin_it, lhs_it),
                         std::distance(begin_it, rhs_it), monkey.value_lhs,
                         monkey.value_rhs,
                         aoc::get_binary_op<value_t>(monkey.operation));
  }
  return {monkeys, root_index, human_index};
}

fn solve_case1(parsed_monkeys_t const& parsed_monkeys) -> i64 {
  auto [monkeys, root_index, human_index] = transform_monkeys(parsed_monkeys);
  return static_cast<i64>(get_yelled_number(monkeys, root_index));
}

fn solve_case2(parsed_monkeys_t const& parsed_monkeys) -> i64 {
  auto [monkeys, root_index, human_index] = transform_monkeys(parsed_monkeys);

  monkey_it begin_it = std::begin(monkeys);
  let begin_parsed_it = std::begin(parsed_monkeys);
  let get_index = [&](auto it) {
    if constexpr (std::same_as<decltype(it), monkey_it>) {
      return std::distance(begin_it, it);
    } else {
      return std::distance(begin_parsed_it, it);
    }
  };

  monkey_it current_it = begin_it + human_index;
  loop {
    let current_index = get_index(current_it);
    str current_name = parsed_monkeys[current_index].first;

    let parsed_parent_it = stdr::find_if(parsed_monkeys, [&](let& monkey_pair) {
      return (monkey_pair.second.index_lhs == current_name) ||
             (monkey_pair.second.index_rhs == current_name);
    });
    let parent_index = get_index(parsed_parent_it);
    monkey_it parent_it = begin_it + parent_index;

    bool current_was_left = (parent_it->index_lhs == current_index);
    let sibling_index =
        current_was_left ? parent_it->index_rhs : parent_it->index_lhs;
    char parent_op = parsed_parent_it->second.operation;

    current_it->value_lhs = None;
    current_it->value_rhs = None;

    if (parent_index == root_index) {
      *parent_it = get_value_monkey(0, 0);
      current_it->index_lhs = parent_index;
      current_it->index_rhs = sibling_index;
      current_it->operation = std::plus{};
      break;
    }

    // For non-commutative ops where current was on the right:
    //   parent = sibling - current -> current = sibling - parent
    //   parent = sibling / current -> current = sibling / parent
    // Swap operand order and use the original op instead of its inverse.
    bool swap_operands =
        !current_was_left && (parent_op == '-' || parent_op == '/');
    current_it->index_lhs = swap_operands ? sibling_index : parent_index;
    current_it->index_rhs = swap_operands ? parent_index : sibling_index;
    current_it->operation =
        swap_operands ? aoc::get_binary_op<value_t>(parent_op)
                      : aoc::get_inverse_binary_op<value_t>(parent_op);

    current_it = parent_it;
  }

  return static_cast<i64>(get_yelled_number(monkeys, human_index));
}

int main() {
  std::println("Part 1");
  let example = parse("day21.example");
  AOC_EXPECT_RESULT(152, solve_case1(example));
  let input = parse("day21.input");
  AOC_EXPECT_RESULT(291425799367130, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(301, solve_case2(example));
  AOC_EXPECT_RESULT(3219579395609, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day21.rs
type Ratio = aoc::math::Ratio<i64>;

#[derive(Clone, Copy, PartialEq, Eq)]
enum Op {
    Value,
    Add,
    Sub,
    Mul,
    Div,
}

impl Op {
    fn apply(self, lhs: Ratio, rhs: Ratio) -> Ratio {
        match self {
            Op::Value | Op::Add => lhs + rhs,
            Op::Sub => lhs - rhs,
            Op::Mul => lhs * rhs,
            Op::Div => lhs / rhs,
        }
    }

    fn inverse(self) -> Self {
        match self {
            Op::Add | Op::Value => Op::Sub,
            Op::Sub => Op::Add,
            Op::Mul => Op::Div,
            Op::Div => Op::Mul,
        }
    }

    fn is_noncommutative(self) -> bool {
        matches!(self, Op::Sub | Op::Div)
    }
}

struct ParsedMonkey {
    name_lhs: String,
    name_rhs: String,
    value_lhs: Option<Ratio>,
    value_rhs: Option<Ratio>,
    op: Op,
}

type ParsedMonkeys = Vec<(String, ParsedMonkey)>;

fn parse(filename: &str) -> ParsedMonkeys {
    aoc::file::read_lines(filename)
        .into_iter()
        .map(|line| {
            let (name, rest) = line.split_once(':').unwrap();
            let parts: Vec<&str> = rest[1..].split(' ').collect();
            let pm = if parts.len() == 1 {
                let value = Ratio::integer(parts[0].parse::<i64>().unwrap());
                ParsedMonkey {
                    name_lhs: String::new(),
                    name_rhs: String::new(),
                    value_lhs: Some(value),
                    value_rhs: Some(Ratio::integer(0)),
                    op: Op::Value,
                }
            } else {
                ParsedMonkey {
                    name_lhs: parts[0].to_string(),
                    name_rhs: parts[2].to_string(),
                    value_lhs: None,
                    value_rhs: None,
                    op: match parts[1] {
                        "+" => Op::Add,
                        "-" => Op::Sub,
                        "*" => Op::Mul,
                        "/" => Op::Div,
                        _ => unreachable!(),
                    },
                }
            };
            (name.to_string(), pm)
        })
        .collect()
}

#[derive(Clone)]
struct Monkey {
    index_lhs: usize,
    index_rhs: usize,
    value_lhs: Option<Ratio>,
    value_rhs: Option<Ratio>,
    op: Op,
}

impl Monkey {
    fn value(lhs: Ratio, rhs: Ratio) -> Self {
        Self {
            index_lhs: usize::MAX,
            index_rhs: usize::MAX,
            value_lhs: Some(lhs),
            value_rhs: Some(rhs),
            op: Op::Add,
        }
    }

    fn is_ready(&self) -> bool {
        self.value_lhs.is_some() && self.value_rhs.is_some()
    }

    fn apply(&self) -> Ratio {
        self.op
            .apply(self.value_lhs.unwrap(), self.value_rhs.unwrap())
    }
}

fn transform_monkeys(parsed: &ParsedMonkeys) -> (Vec<Monkey>, usize, usize) {
    let mut monkeys = Vec::new();
    let mut root_index = 0;
    let mut human_index = 0;
    for (index, (name, pm)) in parsed.iter().enumerate() {
        if name == "root" {
            root_index = index;
        }
        if name == "humn" {
            human_index = index;
        }

        if pm.op == Op::Value {
            monkeys.push(Monkey::value(pm.value_lhs.unwrap(), pm.value_rhs.unwrap()));
            continue;
        }

        let index_lhs = parsed.iter().position(|(n, _)| n == &pm.name_lhs).unwrap();
        let index_rhs = parsed.iter().position(|(n, _)| n == &pm.name_rhs).unwrap();
        monkeys.push(Monkey {
            index_lhs,
            index_rhs,
            value_lhs: pm.value_lhs,
            value_rhs: pm.value_rhs,
            op: pm.op,
        });
    }
    (monkeys, root_index, human_index)
}

fn get_yelled_number(monkeys: &mut Vec<Monkey>, current: usize) -> Ratio {
    if monkeys[current].is_ready() {
        return monkeys[current].apply();
    }
    let index_lhs = monkeys[current].index_lhs;
    let index_rhs = monkeys[current].index_rhs;
    let lhs = get_yelled_number(monkeys, index_lhs);
    let rhs = get_yelled_number(monkeys, index_rhs);
    monkeys[current].value_lhs = Some(lhs);
    monkeys[current].value_rhs = Some(rhs);
    monkeys[current].apply()
}

fn solve_case1(parsed: &ParsedMonkeys) -> i64 {
    let (mut monkeys, root_index, _) = transform_monkeys(parsed);
    get_yelled_number(&mut monkeys, root_index).as_integer()
}

fn solve_case2(parsed: &ParsedMonkeys) -> i64 {
    let (mut monkeys, root_index, human_index) = transform_monkeys(parsed);

    let mut current_index = human_index;
    loop {
        let current_name = &parsed[current_index].0;
        let parent_index = parsed
            .iter()
            .position(|(_, pm)| pm.name_lhs == *current_name || pm.name_rhs == *current_name)
            .unwrap();

        let parent_op = monkeys[parent_index].op;
        let current_was_left = monkeys[parent_index].index_lhs == current_index;
        let sibling_index = if current_was_left {
            monkeys[parent_index].index_rhs
        } else {
            monkeys[parent_index].index_lhs
        };

        monkeys[current_index].value_lhs = None;
        monkeys[current_index].value_rhs = None;

        if parent_index == root_index {
            monkeys[parent_index] = Monkey::value(Ratio::integer(0), Ratio::integer(0));
            monkeys[current_index].index_lhs = parent_index;
            monkeys[current_index].index_rhs = sibling_index;
            monkeys[current_index].op = Op::Add;
            break;
        }

        // For non-commutative ops where current was on the right:
        //   parent = sibling - current -> current = sibling - parent
        //   parent = sibling / current -> current = sibling / parent
        // Swap operand order and use the original op instead of its inverse
        let swap_operands = !current_was_left && parent_op.is_noncommutative();
        monkeys[current_index].index_lhs = if swap_operands {
            sibling_index
        } else {
            parent_index
        };
        monkeys[current_index].index_rhs = if swap_operands {
            parent_index
        } else {
            sibling_index
        };
        monkeys[current_index].op = if swap_operands {
            parent_op
        } else {
            parent_op.inverse()
        };

        current_index = parent_index;
    }

    get_yelled_number(&mut monkeys, human_index).as_integer()
}

fn main() {
    println!("Part 1");
    let example = parse("day21.example");
    aoc::expect_result!(152, solve_case1(&example));
    let input = parse("day21.input");
    aoc::expect_result!(291425799367130_i64, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(301, solve_case2(&example));
    aoc::expect_result!(3219579395609_i64, solve_case2(&input));
}
